client/beefy/src/round.rs

```diff
@@ -53,88 +57,180 @@ fn threshold(authorities: usize) -> usize {
+		if block_number > self.session_start {
+			self.validator_set.id()
+		} else {
+			self.prev_validator_set.id()
+		}
 	}
 
-	pub(crate) fn validators(&self) -> &[Public] {
-		self.validator_set.validators()
+	pub(crate) fn validators_for(&self, block_number: NumberFor<B>) -> &[Public] {
+		if block_number > self.session_start {
+			self.validator_set.validators()
+		} else {
+			self.prev_validator_set.validators()
+		}
 	}
 
-	pub(crate) fn add_vote(&mut self, round: &(H, N), vote: (Public, Signature)) -> bool {
-		if self.validator_set.validators().iter().any(|id| vote.0 == *id) {
-			self.rounds.entry(round.clone()).or_default().add_vote(vote)
-		} else {
+	pub(crate) fn validator_set(&self) -> &ValidatorSet<Public> {
+		&self.validator_set
+	}
+
+	pub(crate) fn session_start(&self) -> &NumberFor<B> {
+		&self.session_start
+	}
+
+	pub(crate) fn should_self_vote(&self, round: &(P, NumberFor<B>)) -> bool {
+		Some(round.1.clone()) > self.best_done &&
+			self.rounds.get(round).map(|tracker| !tracker.has_self_vote()).unwrap_or(true)
+	}
+
+	pub(crate) fn add_vote(
+		&mut self,
+		round: &(P, NumberFor<B>),
+		vote: (Public, Signature),
+		self_vote: bool,
+	) -> bool {
+		if Some(round.1.clone()) <= self.best_done {
+			debug!(
+				target: "beefy",
+				"🥩 received vote for old stale round {:?}, ignoring",
+				round.1
+			);
 			false
+		} else if !self.validator_set.validators().iter().any(|id| vote.0 == *id) {
+			debug!(
+				target: "beefy",
+				"🥩 received vote {:?} from validator that is not in the validator set, ignoring",
+				vote
+			);
+			false
+		} else {
+			self.rounds.entry(round.clone()).or_default().add_vote(vote, self_vote)
 		}
 	}
 
-	pub(crate) fn is_done(&self, round: &(H, N)) -> bool {
+	pub(crate) fn try_conclude(
+		&mut self,
+		round: &(P, NumberFor<B>),
+	) -> Option<Vec<Option<Signature>>> {
 		let done = self
 			.rounds
 			.get(round)
 			.map(|tracker| tracker.is_done(threshold(self.validator_set.len())))
 			.unwrap_or(false);
-
-		debug!(target: "beefy", "🥩 Round #{} done: {}", round.1, done);
-
-		done
-	}
-
-	pub(crate) fn drop(&mut self, round: &(H, N)) -> Option<Vec<Option<Signature>>> {
-		trace!(target: "beefy", "🥩 About to drop round #{}", round.1);
-
-		let signatures = self.rounds.remove(round)?.votes;
-
-		Some(
-			self.validator_set
-				.validators()
-				.iter()
-				.map(|authority_id| {
-					signatures.iter().find_map(|(id, sig)| {
-						if id == authority_id {
-							Some(sig.clone())
-						} else {
-							None
-						}
+		trace!(target: "beefy", "🥩 Round #{} done: {}", round.1, done);
+
+		if done {
+			let signatures = self.rounds.remove(round)?.votes;
+			self.best_done = self.best_done.clone().max(Some(round.1.clone()));
+			trace!(target: "beefy", "🥩 Concluded round #{}", round.1);
+
+			Some(
+				self.validator_set
+					.validators()
+					.iter()
+					.map(|authority_id| {
+						signatures.iter().find_map(|(id, sig)| {
```

Comment:
I was thinking to suggest change type of `RoundTracker::votes` to some map to avoid this loop - iiuc the number of BEEFY validators = number of GRANDPA validators, which is e.g. ~1K on Kusama && this loop is `N^2`. But now I'm unsure if that's correct.

So I have a question - what if the same validator sends two different signatures. I don't see any checks in `Rounds::add_vote` - is it handled somewhere else? What I'm worrying about is that the `RoundTracker::votes` will have two different entries for the same validator. And then in `try_conclude` we are just looking whether `votes.len() >= threshold`. Is it possible? Sorry for possibly being dumb here - I'm looking at this code for the first time :)

Reply:
done